File: go/28_Find_the_Index_of_the_First_Occurrence_in_a_String/28_Find_the_Index_of_the_First_Occurrence_in_a_String.go
func strStr(haystack string, needle string) int {
    // Rabin Karp
    n := len(haystack)
    m := len(needle)
    for i:= 0; i < n-m+1; i++ {
        if haystack[i:i+m] == needle {
            return i
        }
    }
    return -1
}

File: java/28_Find_the_Index_of_the_First_Occurrence_in_a_String/28_Find_the_Index_of_the_First_Occurrence_in_a_String.java
class Solution {
    public int strStr(String haystack, String needle) {
        int sourceLength = haystack.length();
        int targetLength = needle.length();

        if (targetLength > sourceLength) {
            return -1;
        }

        if (sourceLength == 0) {
            return -1;
        } else if (targetLength == 0) {
            return 0;
        }

        for (int i = 0; i < sourceLength; i++) {
            if (i + targetLength > sourceLength) {
                break;
            }
            if (haystack.charAt(i) == needle.charAt(0)) {
                for (int j = 0; j < targetLength; j++) {
                    if (needle.charAt(j) != haystack.charAt(i+j)) {
                        break;
                    } else if (j == targetLength - 1) {
                        return i;
                    }
                }
            }
        }
        return -1;
    }
}
